activerecord/CHANGELOG.md

```diff
@@ -1,3 +1,22 @@
+*   Association#reset also resets the cached `association_scope`
+
+    After loading a `has_one` association on an object then resetting the
+    association, calling save on the object has the effect of loading the
+    association again (as part of
+    `AutosaveAssocation#save_has_one_association`). When the association is
+    loaded in this way, the query to find the associated record was based on
+    the `association_scope` as evaluated originally. The conditions were
+    not _re_-evaluated, meaning they could be stale. If the scope includes
+    `Time.current`, for example, the cached `association_scope` would include
+    an old timestamp.
+
+    Now, when resetting an association, the `association_scope` is also reset.
+    hen the association is later reloaded, the scope is evaluated afresh.
```

Comment:
This could be a bit more succinct. Here's my understanding, feel free to edit if I'm wrong.

```suggestion
    `Association#reset` also resets the association scopes.

    After loading a `has_one` association on an object and calling `reset`,
    calling `save` could incorrectly cache scopes, which could result in
    incorrect associated objects.
    
    Now when resetting an association, the association's scope is also reset.
```

Reply:
Thanks. I've addressed this [here](https://github.com/rails/rails/pull/42643/commits/e33d9a67c2fdf5ca5b32724d1ea7b7e1e9f2e351).